### ISSUE TITLE
Network locking is defective and has other issues

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#380]
Originally reported of Google Code with ID 324.

```
Below is a list of issues I found in LUNetwork* and related code during a quick review. There might be more.

General:
- Why are so many errors.AddressPoolError swallowed silently?
- Exception handling should be more targeted, e.g. try: pool.Reserve(…); except …: pass; else: logging.info(…)

LUNetworkAdd:
- Uses nodes without locking
- CheckPrereq: MAC address check can go into CheckArguments, should use custom error message as the second half is hardcoded
Exec uses hardcoded IP address family (“4”)

LUNetworkRemove:
- Uses node groups without locking

LUNetworkSetParams:
- ExpandNames: Gets network from config before actually having the lock
- CheckPrereq: Same issue as LUNetworkAdd regarding MAC address

_NetworkQuery:
- Doesn't support synchronous queries
- Duplicate logic from _GroupQuery, should be abstracted

LUNetworkConnect:
- Doesn't lock network at all
- ExpandNames: Gets network and node group from config before actually having the lock
- Doesn't lock node group
- Acquired instance locks are not verified (even though there's a copy-and-pasted comment saying so)
- CheckPrereq: Uses cfg.GetInstanceInfo many times instead of one GetMultiInstanceInfo
- Exec: Should use owned locks instead of calling cfg.GetNodeGroupInstances again

LUNetworkDisconnect:
- Doesn't lock network at all
- ExpandNames: Gets network and node group from config before actually having the lock
- Exec: Should use owned locks instead of calling cfg.GetNodeGroupInstances again

lib/network.py:
- Lacks many docstrings

```
Originally added on 2012-11-29 12:55:56 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Michael,

IIRC there have been a lot of patches related to this. How is it going? Should the issue be closed, or at least updated with what's done and what's missing?

Thanks a lot,

Guido

```
Originally added on 2013-02-11 15:39:17 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
There are still some issues left (and I may have missed some).

LUNetworkConnect:
- Acquired instance locks are not verified (even though there's a copy-and-pasted comment saying so)

LUNetworkDisconnect:
- Acquired instance locks are not verified (even though there's a copy-and-pasted comment saying so)

lib/network.py:
- Lacks many docstrings

In addition there are plenty of early returns (“if not foo: return”). Those can be confusing and easy to miss when used in the middle of CheckPrereq/Exec. They should be replaced with proper if/else branches.
```
Originally added on 2013-02-12 11:29:54 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Dimitris,

When you have time could you take a look at this issue as well?

Thanks a lot,
Guido

```
Originally added on 2013-02-12 11:42:57 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi!

Sorry for the delay. As I see instance locks are indeed checked in _NetworkConflictCheck(). I agree with the doc strings and the if/else branches
and I will fix them, but I would not consider them critical, right?

Thanks,
dimara


```
Originally added on 2013-02-18 16:38:37 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Please move the lock check to the beginning of CheckPrereq. You don't want to proceed with anything if the acquired locks are incorrect (likely due to a bug somewhere).
```
Originally added on 2013-02-19 14:39:18 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
_NetworkConflictCheck() which invokes _CheckNodeGroupInstances() is already in CheckPrereq. Moving it further up will not make a difference because both opcodes take no actions in CheckPrereq. Do you want me still to move this a few lines above?
```
Originally added on 2013-02-19 16:50:38 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Yes, and it should not depend on “self.op.conflicts_check” as it currently does in some opcodes (NetworkConnect at least). You may not acquire locks without conflict checks, but the locks should still be verified (even if there are none).
```
Originally added on 2013-02-19 16:55:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Is this now fixed, with the latest patches?

Thanks,

Guido
```
Originally added on 2013-02-26 16:05:20 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
No, I still see early returns.
```
Originally added on 2013-02-26 16:14:34 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

what do you mean exactly? What should be fixed?

Thanks,
dimara
```
Originally added on 2013-02-27 10:47:33 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Let me rephrase: None of the other LUs use early returns in the middle of CheckPrereq or Exec. They make the code harder to follow. There are exceptions, such as some utility functions starting with code like “if not param: return False”, but in general early returns in the main body of a function should be avoided.
```
Originally added on 2013-02-27 12:41:37 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This is finally fixed, with:
ce35bde Remove early returns in network LUs

```
Originally added on 2013-03-01 19:33:35 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
